File: packages/textprompts-ts/src/savers.ts
import { writeFile } from "node:fs/promises";

import YAML from "yaml";

import { TextPromptsError } from "./errors";
import type { FlagDecl, VarDecl } from "./frontmatter-schema";
import { Prompt, type PromptMeta } from "./models";

export type FrontMatterFormat = "toml" | "yaml";

const escapeTomlString = (value: string): string =>
  value.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n").replace(/\r/g, "\\r");

const tomlString = (value: string | null | undefined): string =>
  `"${escapeTomlString(value ?? "")}"`;

const tomlValue = (value: unknown): string | null => {
  if (value == null) return null;
  if (typeof value === "string") return tomlString(value);
  if (typeof value === "boolean") return String(value);
  if (typeof value === "number") return String(value);
  if (Array.isArray(value)) {
    if (value.every((v) => typeof v === "string")) {
      return `[${value.map((v) => tomlString(v as string)).join(", ")}]`;
    }
    if (value.every((v) => typeof v === "number")) {
      return `[${(value as number[]).map(String).join(", ")}]`;
    }
    if (value.every((v) => typeof v === "boolean")) {
      return `[${(value as boolean[]).map(String).join(", ")}]`;
    }
    return null;
  }
  return null;
};

const tomlInlineTable = (value: Record<string, unknown>): string | null => {
  const parts: string[] = [];
  for (const [k, v] of Object.entries(value)) {
    const serialized = tomlValue(v);
    if (serialized === null) return null;
    parts.push(`${k} = ${serialized}`);
  }
  return `{ ${parts.join(", ")} }`;
};

const isPlainObject = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" &&
  v !== null &&
  !Array.isArray(v) &&
  !(v instanceof Date) &&
  Object.getPrototypeOf(v) === Object.prototype;

// ---------------------------------------------------------------------------
// TOML serializer for the full v2 meta object.
// ---------------------------------------------------------------------------

const serializeTomlExtraLine = (key: string, value: unknown): string | null => {
  if (value == null) return null;
  if (typeof value === "string") return `${key} = ${tomlString(value)}`;
  if (typeof value === "boolean") return `${key} = ${String(value)}`;
  if (typeof value === "number") return `${key} = ${String(value)}`;
  if (Array.isArray(value)) {
    const v = tomlValue(value);
    return v === null ? null : `${key} = ${v}`;
  }
  if (isPlainObject(value)) {
    const inline = tomlInlineTable(value);
    return inline === null ? null : `${key} = ${inline}`;
  }
  return null;
};

const serializeTomlFlag = (name: string, decl: FlagDecl): string[] => {
  const lines: string[] = [`[flags.${name}]`];
  // type: always emit explicit for clarity except boolean default is allowed
  if (decl.kind === "enum") {
    lines.push(`type = "enum"`);
    lines.push(`values = ${tomlValue(decl.values)}`);
  } else {
    lines.push(`type = "boolean"`);
  }
  if (decl.description !== undefined) {
    lines.push(`description = ${tomlString(decl.description)}`);
  }
  for (const [k, v] of Object.entries(decl.extras)) {
    const line = serializeTomlExtraLine(k, v);
    if (line === null) {
      throw new TextPromptsError(
        `Cannot serialize extras key '${k}' of flag '${name}' to TOML: unsupported value`,
      );
    }
    lines.push(line);
  }
  return lines;
};

const serializeTomlVariable = (name: string, decl: VarDecl): string[] => {
  const lines: string[] = [`[variables.${name}]`];
  if (decl.description !== undefined) {
    lines.push(`description = ${tomlString(decl.description)}`);
  }
  for (const [k, v] of Object.entries(decl.extras)) {
    const line = serializeTomlExtraLine(k, v);
    if (line === null) {
      throw new TextPromptsError(
        `Cannot serialize extras key '${k}' of variable '${name}' to TOML: unsupported value`,
      );
    }
    lines.push(line);
  }
  return lines;
};

const serializeTomlMeta = (meta: PromptMeta): string => {
  const lines: string[] = ["---"];
  lines.push(`title = ${tomlString(meta.title)}`);
  lines.push(`description = ${tomlString(meta.description)}`);
  lines.push(`version = ${tomlString(meta.version)}`);
  if (meta.author) lines.push(`author = ${tomlString(meta.author)}`);
  if (meta.created) lines.push(`created = ${tomlString(meta.created)}`);

  for (const [key, value] of Object.entries(meta.extras)) {
    const line = serializeTomlExtraLine(key, value);
    if (line === null) {
      throw new TextPromptsError(
        `Cannot serialize top-level extras key '${key}' to TOML: unsupported value`,
      );
    }
    lines.push(line);
  }

  const flagNames = Object.keys(meta.flags);
  for (const name of flagNames) {
    lines.push("");
    const decl = meta.flags[name];
    if (decl === undefined) continue;
    lines.push(...serializeTomlFlag(name, decl));
  }

  const varNames = Object.keys(meta.variables);
  for (const name of varNames) {
    lines.push("");
    const decl = meta.variables[name];
    if (decl === undefined) continue;
    lines.push(...serializeTomlVariable(name, decl));
  }

  lines.push("---");
  return lines.join("\n");
};

// ---------------------------------------------------------------------------
// YAML serializer (uses `yaml` package for complex values).
// ---------------------------------------------------------------------------

const serializeYamlObject = (obj: Record<string, unknown>): string => {
  // YAML.stringify emits a trailing newline; strip it so we can join manually.
  return YAML.stringify(obj).trimEnd();
};

const serializeYamlMeta = (meta: PromptMeta): string => {
  const root: Record<string, unknown> = {};
  if (meta.title != null) root.title = meta.title;
  if (meta.description != null) root.description = meta.description;
  if (meta.version != null) root.version = meta.version;
  if (meta.author != null) root.author = meta.author;
  if (meta.created != null) root.created = meta.created;
  for (const [k, v] of Object.entries(meta.extras)) {
    root[k] = v;
  }

  const flagNames = Object.keys(meta.flags);
  if (flagNames.length > 0) {
    const flagsObj: Record<string, unknown> = {};
    for (const name of flagNames) {
      const decl = meta.flags[name];
      if (decl === undefined) continue;
      const entry: Record<string, unknown> = {};
      entry.type = decl.kind;
      if (decl.kind === "enum") entry.values = decl.values;
      if (decl.description !== undefined) entry.description = decl.description;
      for (const [k, v] of Object.entries(decl.extras)) entry[k] = v;
      flagsObj[name] = entry;
    }
    root.flags = flagsObj;
  }

  const varNames = Object.keys(meta.variables);
  if (varNames.length > 0) {
    const varsObj: Record<string, unknown> = {};
    for (const name of varNames) {
      const decl = meta.variables[name];
      if (decl === undefined) continue;
      const entry: Record<string, unknown> = {};
      if (decl.description !== undefined) entry.description = decl.description;
      for (const [k, v] of Object.entries(decl.extras)) entry[k] = v;
      varsObj[name] = entry;
    }
    root.variables = varsObj;
  }

  // Fall back: emit minimal title/description/version even when nothing is set
  // (matches v1 saver behaviour and existing tests).
  if (!("title" in root)) root.title = "";
  if (!("description" in root)) root.description = "";
  if (!("version" in root)) root.version = "";

  const body = serializeYamlObject(root);
  return `---\n${body}\n---`;
};

// ---------------------------------------------------------------------------
// Public API.
// ---------------------------------------------------------------------------

export const savePrompt = async (
  path: string,
  content: string | Prompt,
  options?: { format?: FrontMatterFormat },
): Promise<void> => {
  const format = options?.format ?? "toml";

  if (typeof content === "string") {
    const template =
      format === "yaml"
        ? `---\ntitle: ""\ndescription: ""\nversion: ""\n---\n\n${content}`
        : `---\ntitle = ""\ndescription = ""\nversion = ""\n---\n\n${content}`;
    await writeFile(path, template, { encoding: "utf8" });
    return;
  }

  if (!(content instanceof Prompt)) {
    throw new TypeError(`content must be string or Prompt, received ${typeof content}`);
  }

  const meta: PromptMeta = content.meta ?? { extras: {}, flags: {}, variables: {} };

  // For YAML we need a meta with at least title/description/version represented
  // even if null, because parseString expects them as strings on round-trip
  // when no override is given. The existing behavior emits empty strings.
  const normalized: PromptMeta = {
    ...meta,
    title: meta.title ?? "",
    description: meta.description ?? "",
    version: meta.version ?? "",
  };

  let serialized: string;
  if (format === "yaml") {
    serialized = serializeYamlMeta(normalized);
  } else {
    serialized = serializeTomlMeta(normalized);
  }

  const body = content.prompt.toString();
  await writeFile(path, `${serialized}\n\n${body}`, { encoding: "utf8" });
};

File: src/textprompts/savers.py
"""Prompt file serializer (SPEC §6, save side).

Emits a Prompt's metadata and body back to disk. Both TOML and YAML output
paths round-trip the full v2 schema: standard fields, top-level extras,
``[flags.*]`` declarations, ``[variables.*]`` declarations, and per-decl
extras. ``meta.extras`` MUST NOT contain raw ``flags``/``variables`` keys —
the loader routes those to ``meta.flags`` / ``meta.variables`` already.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any, Literal, Union

import yaml as yaml_lib  # type: ignore[import-untyped]

from .errors import TextPromptsError
from .models import FlagDecl, Prompt, PromptMeta, VariableDecl

# ---------------------------------------------------------------------------
# TOML primitives.
# ---------------------------------------------------------------------------


def _escape_toml(value: str) -> str:
    """Escape a string for TOML output (for use inside double quotes)."""
    if not value:
        return ""
    return (
        value.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )


def _toml_string(value: Union[str, None]) -> str:
    return f'"{_escape_toml(value or "")}"'


def _toml_value(value: Any) -> Union[str, None]:
    """Serialize a primitive TOML value, or return None if unsupported."""
    if value is None:
        return None  # TOML has no null
    if isinstance(value, bool):  # bool must come before int check
        return str(value).lower()
    if isinstance(value, str):
        return _toml_string(value)
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, (list, tuple)):
        items: list[str] = []
        for v in value:
            s = _toml_value(v)
            if s is None:
                return None
            items.append(s)
        return f"[{', '.join(items)}]"
    return None  # nested objects, dates, etc.


def _toml_extra_line(key: str, value: Any) -> Union[str, None]:
    """Serialize a single extras key/value as a top-level TOML line."""
    serialized = _toml_value(value)
    if serialized is None:
        # Fallback: try inline table for plain dicts.
        if isinstance(value, dict):
            parts: list[str] = []
            for k, v in value.items():
                s = _toml_value(v)
                if s is None:
                    return None
                parts.append(f"{k} = {s}")
            return f"{key} = {{ {', '.join(parts)} }}"
        return None
    return f"{key} = {serialized}"


def _toml_flag_lines(name: str, decl: FlagDecl) -> list[str]:
    lines: list[str] = [f"[flags.{name}]"]
    if decl.kind == "enum":
        lines.append('type = "enum"')
        values_serialized = _toml_value(list(decl.values or ()))
        lines.append(f"values = {values_serialized}")
    else:
        lines.append('type = "boolean"')
    if decl.description is not None:
        lines.append(f"description = {_toml_string(decl.description)}")
    for k, v in decl.extras.items():
        line = _toml_extra_line(k, v)
        if line is None:
            raise TextPromptsError(
                f"Cannot serialize extras key '{k}' of flag '{name}' to TOML: "
                f"unsupported value type {type(v).__name__}"
            )
        lines.append(line)
    return lines


def _toml_variable_lines(name: str, decl: VariableDecl) -> list[str]:
    lines: list[str] = [f"[variables.{name}]"]
    if decl.description is not None:
        lines.append(f"description = {_toml_string(decl.description)}")
    for k, v in decl.extras.items():
        line = _toml_extra_line(k, v)
        if line is None:
            raise TextPromptsError(
                f"Cannot serialize extras key '{k}' of variable '{name}' to TOML: "
                f"unsupported value type {type(v).__name__}"
            )
        lines.append(line)
    return lines


def _serialize_toml_meta(meta: PromptMeta) -> str:
    lines: list[str] = ["---"]
    lines.append(f"title = {_toml_string(meta.title)}")
    lines.append(f"description = {_toml_string(meta.description)}")
    lines.append(f"version = {_toml_string(meta.version)}")
    if meta.author:
        lines.append(f"author = {_toml_string(meta.author)}")
    if meta.created:
        lines.append(f"created = {_toml_string(meta.created.isoformat())}")
    for key, value in meta.extras.items():
        if key in {"flags", "variables"}:
            # Should never happen: loader routes these to typed fields.
            continue
        line = _toml_extra_line(key, value)
        if line is None:
            raise TextPromptsError(
                f"Cannot serialize top-level extras key '{key}' to TOML: "
                f"unsupported value type {type(value).__name__}"
            )
        lines.append(line)

    for flag_name, flag_decl in meta.flags.items():
        lines.append("")
        lines.extend(_toml_flag_lines(flag_name, flag_decl))

    for var_name, var_decl in meta.variables.items():
        lines.append("")
        lines.extend(_toml_variable_lines(var_name, var_decl))

    lines.append("---")
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# YAML.
# ---------------------------------------------------------------------------


def _serialize_yaml_meta(meta: PromptMeta) -> str:
    root: dict[str, Any] = {}
    # Always emit title/description/version (even when empty) to preserve v1
    # round-trip behavior.
    root["title"] = meta.title or ""
    root["description"] = meta.description or ""
    root["version"] = meta.version or ""
    if meta.author is not None:
        root["author"] = meta.author
    if meta.created is not None:
        root["created"] = meta.created.isoformat()
    for k, v in meta.extras.items():
        if k in {"flags", "variables"}:
            continue
        root[k] = v

    if meta.flags:
        flags_obj: dict[str, Any] = {}
        for flag_name, flag_decl in meta.flags.items():
            entry: dict[str, Any] = {"type": flag_decl.kind}
            if flag_decl.kind == "enum":
                entry["values"] = list(flag_decl.values or ())
            if flag_decl.description is not None:
                entry["description"] = flag_decl.description
            for k, v in flag_decl.extras.items():
                entry[k] = v
            flags_obj[flag_name] = entry
        root["flags"] = flags_obj

    if meta.variables:
        vars_obj: dict[str, Any] = {}
        for var_name, var_decl in meta.variables.items():
            var_entry: dict[str, Any] = {}
            if var_decl.description is not None:
                var_entry["description"] = var_decl.description
            for k, v in var_decl.extras.items():
                var_entry[k] = v
            vars_obj[var_name] = var_entry
        root["variables"] = vars_obj

    body = yaml_lib.dump(root, default_flow_style=False, sort_keys=False).rstrip("\n")
    return f"---\n{body}\n---"


# ---------------------------------------------------------------------------
# Public API.
# ---------------------------------------------------------------------------


def save_prompt(
    path: Union[str, Path],
    content: Union[str, Prompt],
    *,
    format: Literal["toml", "yaml"] = "toml",
) -> None:
    """
    Save a prompt to a file.

    Parameters
    ----------
    path : str | Path
        File path to save the prompt to.
    content : str | Prompt
        Either a raw string body or a Prompt with full v2 metadata.
    format : "toml" | "yaml", default "toml"
        Frontmatter format.

    Examples
    --------
    >>> save_prompt("my_prompt.txt", "You are a helpful assistant.")
    >>> save_prompt("my_prompt.txt", "You are a helpful assistant.", format="yaml")
    >>> save_prompt("my_prompt.txt", prompt_obj, format="yaml")
    """
    path = Path(path)

    if isinstance(content, str):
        if format == "yaml":
            template = f'---\ntitle: ""\ndescription: ""\nversion: ""\n---\n\n{content}'
        else:
            template = (
                f'---\ntitle = ""\ndescription = ""\nversion = ""\n---\n\n{content}'
            )
        path.write_text(template, encoding="utf-8")
        return

    if not isinstance(content, Prompt):
        raise TypeError(f"content must be str or Prompt, not {type(content).__name__}")

    meta = content.meta or PromptMeta()
    if format == "yaml":
        serialized = _serialize_yaml_meta(meta)
    else:
        serialized = _serialize_toml_meta(meta)

    body = str(content.prompt)
    path.write_text(f"{serialized}\n\n{body}", encoding="utf-8")
